docs/proposals/volume-provisioning.md

```diff
@@ -0,0 +1,315 @@
+
+1.  Storage classes should be able to hold a set of provisioner parameters
+2.  The provisioner should be passed the claim and the details of the storage
+    class it is provisioning for
+
+#### Out-of-tree provisioners
+
+One of our goals is to enable administrators to create out-of-tree
+provisioners, that is, provisioners whose code does not live in the Kubernetes
+project.  Our experience since the 1.2 release with dynamic provisioning has
+shown that it is impossible to anticipate every aspect and manner of
+provisioning that administrators will want to perform.  The proposed design
+should not prevent future work to allow out-of-tree provisioners.
+
+## Analysis
+
+### Summary of alpha functionality
+
+The alpha functionality for dynamic provisioning works as follows:
+
+1.  A cluster has only a single provisioner at a time, dictated by the cloud
+    provider
+2.  When a new claim is submitted, the controller attempts to find an existing
+    volume that will fulfill the claim.  If a suitable volume is found, the
+    controller binds the claim and volume together and carries on normally.
+3.  If no volume is found for the claim, and the claim has the annotation
+    `volume.alpha.kubernetes.io/storage-class`, the provisioner is invoked
+    to provision a volume to fill that claim
+4.  All provisioners are in-tree; they implement an interface called
+    `ProvisionableVolumePlugin` which has a method, `NewProvisioner()`,
+    that returns a new `Provisioner` instance
+5.  The controller calls the provisioner's `Provision()` method; `Provision`
+    is responsible for provisioning the volume in the cloud provider and
+    returns an `api.PersistentVolume` and an error
+6.  If `Provision` returns an error, the controller stops trying to provision
+7.  If `Provision` returns no error, the controller creates the returned
+    `api.PersistentVolume`, already bound to the claim
+  1.  If the create operation fails, it is retried
+  2.  If the create operation never succeeds, the controller attempts to delete
+      the provisioned volume and creates an event on the claim
+
+### When should we provision volumes?
+
+It is sensible to provision volumes only when an existing volume cannot
+satisfy a claim.  This ensures that existing, already allocated resources are
+used before additional resources are allocated.  In this regard, the existing
+logic need not change much when storage classes are added.  The primary
+difference is that in the existing design, there is no way to parameterize
+provisioners or have more than one provisioner per cluster.  Therefore, once
+we know that we have to provision a volume to satisfy a claim, we need to
+determine which storage class to use for the claim.
+
+### How should a claim be matched to a storage class?
+
+We've discussed so far that volume claims and storage classes will be matched
+using the claim's label selector.  With that said, there are a number of small
```

Comment:
> let's make a specific field for it, say PVC.Spec.RequestedZone

No.  Topology is arbitrary, we should not be promoting one sense of it above others.

> Or generic PVC.Spec.RequestedParameters

This puts you farther away from your own stated goal that users should not know the difference between auto-provisioned and static volumes.  We are already using a selector to choose potential volumes, so now a user would have to fill in both fields, if they want to not care, or else choose which provisioning mode they want.  Why?

> I do not like leaking cloud details (exact availability zone names) into PVCs

Zones affect EVERYTHING.  Most users won't care, and they will leave it blank, but some will care.

> We could introduce zones as 1st class objects:

We could.  That is well beyond the discussion at hand, though.  It also comes with a lot of baggage - what happens when new zones are available in GCE or AWS?  Cluster admins have to create zone objects?  Or we do it automatically and ... name them the same as the cloud provider?

Availability zones are there for people who need them, and should be transparent for people who are not paying attention to them.  I understand the awkwardness of using a selector as "parameters", which is why I keep saying that they are constraints.

A provisioner is taksed with creating a volume that is fit for use by the consumer.  If the consumer has applied any labels, the provisioned needs to respect those labels.  Otherwise it produces volumes that can not be used by the intended consumer.  So either we say that `PVC.spec.selector` is totally ignored when auto-provisioning or we acknowledge that it defines constraints on the provisioner.